[Transparent ascription step 1] Feature split

clementblaudeau opened the pull request:
This PR corresponds to the first step described in [the transparent ascription RFC](https://github.com/ocaml/RFCs/pull/54): the **feature split** between static and dynamic aliases. The nodes for dynamic aliases are called *transparent* as they will eventually also hold a signature information. For now, they only are dynamic aliases.  

# Overview

This PR does the following: 

- [X] Split aliases into static aliases (absent ones) and dynamic aliases  (present ones) in the internal representation of the typechecker.
- [x] Provide attributes (static_alias, dynamic_alias) used by inference
- [X] Remove the presence flag (which coincides exactly with being a static  alias).
- [X] Have dynamic aliases be a subtype of static aliases (a dynamic alias can  be coerced into a static one, provided that the static alias was wellformed).

Crucially, this PR:
- Does not change yet the "aliasable" criterion, i.e. keep the restriction that dynamic aliases cannot contain functor arameter, functor application, etc.
- Maintains the the behavior of `-no-alias-deps`.

# Examples

<details> 
<summary>Dynamic aliases can be obtained directly via attributes:</summary>

```ocaml
module X0 = struct end 
module X1 = X0 [@@dynamic_alias]
```
```ocaml
module X0 : sig end 
module X1 = X0 [@@dynamic_alias]
```
</details>

<details> 
<summary>Strengthening introduces dynamic aliases:</summary>

```ocaml
module M = struct module X = struct end end 
module M' = struct include M end 
```
```ocaml
module M : sig module X : sig end end 
module M' : sig module X = M.X [@@dynamic_alias] end
```
</details>

<details> 
<summary>Dynamic aliases are a subtype of static ones:</summary>

```ocaml
module X0 = struct end
module TestSub : sig module X1 = X0 end =
  struct module X1 = X0 [@@dynamic_alias] end
```
```ocaml
module X0 : sig end
module TestSub : sig module X1 = X0 end
```
</details>

<details> 
<summary>Static aliases are <b>not</b> a subtype of dynamic ones:</summary>

```ocaml
module X0 = struct end
module TestSub : sig module X1 = X0 [@@dynamic_alias] end =
  struct module X1 = X0 [@@static_alias] end
```
```
module X0 : sig end
Line 3, characters 2-44:
3 |   struct module X1 = X0 [@@static_alias] end
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Signature mismatch:
       Modules do not match:
         sig module X1 = X0 end
       is not included in
         sig module X1 = X0 [@@dynamic_alias]  end
       In module "X1":
       Modules do not match:
         (module X0) [@static_alias]
       is not included in
         (module X0) [@dynamic_alias]
|}]
```
</details>

<details>
<summary>Dynamic aliases can be erased by subtyping</summary>

```ocaml
module X0 = struct type t = A | B end
module TestSub : sig module X1 : sig type t end end =
  struct module X1 = X0 [@dynamic_alias] end
```
```ocaml
module X0 : sig type t = A | B val x : t end
module TestSub : sig module X1 : sig type t end end
```

</details>

<details> 
<summary>Both aliases are used by path normalization </summary>

```ocaml
module X0 = struct type t end 
module X1 = X0 [@@dynamic_alias]
module X2 = X0 
let test_eq : X1.t -> X2.t = fun x -> x 
```
```ocaml
module X0 : sig type t end
module X1 = X0 [@@dynamic_alias] 
module X2 = X0
val test_eq : X1.t -> X2.t = <fun>
```
</details>

# Organisation

To ease review, the PR has been split as much as possible into small commits, with some explanations in the commit description. It is organized as follows:

1. Introduce (restricted) transparent signatures 
    a. Renaming aliases into *static_aliases*
    b. Add transparent signatures in types (+Bootstrap)
2. Activate transparent signatures 
    a. Use attributes to infer dynamic aliases
    b. Extend subtyping to transparent signatures 
    c. Change strengthening to infer dynamic aliases in place of present static aliases
    d. Prevent avoidance from inferring present static aliases 
3. Remove module presence
    a. Remove pattern-matching on presence in favor of pattern matching on static aliases (consumers of presence information)
    b. Remove computation of presence (producers of presence information)
    c. Remove the presence field (+ Bootstrap)

# Details

Notable aspects or choices that might be open to discussion:
1. Terminology: dynamic aliases are named `transparent` in signatures, shapes, etc. This is because they will eventually be full blown *transparent signatures*, i.e. signatures of transparent ascription. However, until then, the name might be confusing.
2. No special syntax is added yet, only attributes
3. ~The parsetree was extended with nodes for static aliases and dynamic aliases, although no syntax is provided. This was usefull for the `out_type.ml` file that prints a parsetree from a typedtree. It could maybe be moved to the step 2 of transparent ascription.~
4. There are no tests for the following items, but I would very much like to add some if someone can help me build them:
   * [ ] The behavior of `-no-alias-deps`
   * [ ] The computation/handling of shapes
   * [ ] The documentation generation through ocaml_doc

clementblaudeau committed:
Add transparent signatures NEEDS BOOTSTRAP

- Add a new signature case [Mty_transparent] that is not supported (nor used)
yet, and that aims at replacing present static aliases

- Add tests files [transparent.ml] and [static_aliases.ml] to host future tests

- The [failwith "TODO: transparent ascription step 1"] are placeholders, removed
by the following commits of #14064

clementblaudeau committed:
BOOTSTRAP

clementblaudeau committed:
Rename module "aliases" into "static_aliases"

- Renamed [Mty_alias] into [Mty_static_alias] in types, typedtree, lazy
signatures, shapes, etc.

- Updated shape tests

Note that the aliasing logic is not changed yet, making "static aliases" still a
mix of static and dynamic aliases, as before.

clementblaudeau committed:
Remove pattern-matching on module presence

Transform pattern matches on presence into pattern matches on whether the
signature is a static alias (always absent) or anything else (always
present). The change affects:
1. lambda/translmod (for compilation), where the order of some pattern-matches
are inverted to first deal with static alias, and then other module types.
2. typing/env (for address computation)

clementblaudeau committed:
Extend subtyping to transparent signatures

1. Dynamic aliases are subtype of static ones

2. For both aliases, the aliasing information can be dropped by subtyping,
resolving to the strengthened underlying signature (for instance, [(=P) <? S']
is true if [P] resolves to some [S] and if [S//P <? S'], where [//] denotes
strengthening)

3. Subtyping with aliases is made shallow, the normalization that jumped
directly to the normal form has been removed. This may affect performance of
typechecking

- Proper subtyping tests are added later, once inference with attributes is
working

clementblaudeau committed:
Infer dynamic aliases when strengthening

This commit removes the inference of **present** static aliases through
strengthening: instead, transparent signatures are inferred.

clementblaudeau committed:
Add attributes for inference of static and dynamic aliases

These attributes are used for inference. The fallback case (when no attributes)
is always to infer a static alias (will be changed latter when introducing
dynamic aliases of arbitrary paths).

- Update test cases that display aliases as module types in errors. Static
aliases as signature items are displayed as ambiguous aliases `module X = P`,
while dynamic ones are displayed with an attribute `module X = P
[@@dynamic_alias]`.

clementblaudeau committed:
Add tests for subtyping and inference via attributes

clementblaudeau committed:
Prevent inference of present static aliases via avoidance

Through avoidance, present static aliases could be inferred. This commit removes
the logic of [nondep_mty_with_presence], instead relying on a match on the
result of avoidance to decide on presence. Therefore, avoidance can still infer
static aliases, but the presence flag always correspond to the static/dynamic
type of alias.

clementblaudeau committed:
Remove module presence in env

The typing environment no longer stores the presence information. A
pattern-match on static aliasing replaces it.

clementblaudeau committed:
Remove presence computation in mtype

Several functions in [mtype.ml] were computing presence information by
pattern-matching on the signature. Those have become useless, has being
[Mp_absent] is now exactly the same as being a static alias

clementblaudeau committed:
Remove module presence NEEDS BOOSTRAP

Octachron commented on file_formats/cmt_format.ml:
```diff
@@ -304,8 +304,10 @@ let iter_on_occurrences
-      | Tmty_alias (path, lid) ->
+      | Tmty_static_alias (path, lid) ->
           f ~namespace:Module mty_env path lid
+      | Tmty_transparent _ ->
```
The meaning of `Tmty_transparent`  is not that clear without context. Maybe `Tmty_ascription`, `Tmty_transparent_ascription`, `Tmty_coerced_alias`?

Octachron commented on ocamldoc/odoc_print.ml:
```diff
@@ -55,7 +55,8 @@ let simpl_module_type ?code t =
       Mty_ident _
-    | Mty_alias _ -> t
+    | Mty_static_alias _ -> t
+    | Mty_transparent _ -> failwith "[feature split] to fix"
```
This should be `t`, since we are eliding a module type.

Octachron commented on ocamldoc/odoc_sig.ml:
```diff
@@ -1409,6 +1409,10 @@ module Analyser =
               | Parsetree.Pmty_alias longident ->
                   Name.from_longident longident.txt
+              | Parsetree.Pmty_static_alias _ ->
+                  failwith "[feature split] to fix"
```
Here, we can use the alias name.

clementblaudeau committed:
Extend post-signature-merge invalid-alias check to dynamic aliases

Until dynamic alias can support any module path, signature merging could
introduce invalid aliases. The wellformedness check is therefore extended to
also look for invalid dynamic aliases.

clementblaudeau committed:
Changelog

clementblaudeau commented:
I've updated the description and marked the PR as ready for review (undraft). Sorry for the rebasing noise.

samsa1 commented on typing/shape_reduce.ml:
```diff
@@ -237,7 +241,8 @@ end) = struct
           return (NStruct mnf)
-      | Alias t -> return (NAlias (delay_reduce env t))
+      | Static_alias t -> return (NStatic_alias (delay_reduce env t))
+      | Transparent _t -> return (NTransparent (delay_reduce env t))
```
The `_t` here really looks like a typo.

clementblaudeau replied:
Indeed, and it is quite unfortunate that there is another `t` in the environment with the right type. Fixed by 58899c3a21

clementblaudeau commented:
I've rebased the PR to remove the unnecessary introduction of nodes in the Parsetree. I've updated the description.

clementblaudeau committed:
BOOTSTRAP

clementblaudeau committed:
Remove printing of module presence when debugging typedtree

clementblaudeau commented:
I've rebase (again) to fix the CI. This PR would fix https://github.com/ocaml/ocaml/issues/13976

clementblaudeau committed:
Refactor the inference of aliases (fix incorrect shape + add tests)

The [Pmod_ident] case of inference is the point where aliases can be introduced
by the typechecker. This commit refactors it to fix a few issues:
1. Alias Attributes ([@@dynamic_alias], [@@static_alias]) on non-aliasable paths
emit an error
2. Dynamic aliases now get the correct Transparent shape
3. The proper [Env.scrape_alias] function is used to remove aliases instead of
an ad-hoc inlining.

clementblaudeau commented:
I've rebased on current trunk (24/08/25)

OlivierNicole commented on testsuite/tests/typing-modules/aliases.ml:
```diff
@@ -631,7 +631,10 @@ Line 15, characters 10-30:
 Error: In this "with" constraint, the new definition of "I"
        does not match its original definition in the constrained signature:
-       Modules do not match: (module Int2) is not included in (module Int)
+       Modules do not match:
+         (module Int2) [@static_alias]
+       is not included in
+         (module Int) [@static_alias]
```
Systematically printing whether aliases are static or dynamic in module _types_ can add noise, and is not consistent with the fact that nothing is printed for static module alias declarations. Why not decide that static is the default here too?

OlivierNicole commented on typing/mtype.ml:
```diff
@@ -339,7 +339,7 @@ let rec no_code_needed_mod env pres mty =
       | Mty_functor _ -> false
-      | Mty_alias _ -> false
+      | Mty_static_alias _ -> false
     end
```
I am surprised by this behaviour. I know it’s the same as the current behaviour, but if you know the answer, I’d like to understand: does this mean that a .mli file like:

```ocaml
module M : sig
  type t = int
end

module N = M
```
will require an implementation, even though it possibly doesn’t contain any types? (I guess the answer is yes due to the “possibly” in my sentence.)

OlivierNicole commented on lambda/translmod.ml:
```diff
@@ -275,15 +275,15 @@ let init_shape id modl =
+        init_shape_struct
+          path (Env.add_module_declaration ~check:false
+                             id Mp_absent md env) rem
+    | Sig_module(id, _, md, _, _) :: rem ->
```
Is there a particular reason not to match on `md` directly and use these `when` clauses and predicates instead? It has the drawback of conserving the present/absent terms in the codebase.

Also, is there a particular reason why you are reordering all these patterns?

OlivierNicole commented on lambda/translmod.ml:
```diff
@@ -881,11 +882,9 @@ let rec more_idents = function
+    | Tstr_module { mb_expr={mod_desc = Tmod_structure str}}
     | Tstr_module
-        {mb_presence=Mp_present; mb_expr={mod_desc = Tmod_structure str}}
-    | Tstr_module
-        {mb_presence=Mp_present;
-         mb_expr={mod_desc=
+        { mb_expr={mod_desc=
            Tmod_constraint ({mod_desc = Tmod_structure str}, _, _, _)}} ->
         all_idents str.str_items @ more_idents rem
     | Tstr_module _ -> more_idents rem
```
Doesn’t this change the behaviour? Before the change, only modules of type `Mty_transparent` would add their identifiers to the accumulator.

samsa1 commented on typing/btype.ml:
```diff
@@ -470,7 +470,8 @@ let type_iterators_without_type_expr =
       Mty_ident p
-    | Mty_alias p -> it.it_path p
+    | Mty_static_alias p -> it.it_path p
+    | Mty_transparent p -> it.it_path p
```
Can be slightly cleaned by keeping only one call to `it.it_path` like it was written before.

samsa1 commented on typing/env.ml:
```diff
@@ -1613,7 +1615,15 @@ let rec scrape_alias env ?path mty =
+          (Warnings.No_cmi_file (Path.name path));*)
+        mty
+      end
+  | MtyL_transparent path, _ ->
```
Seems like code duplication between the `MtyL_static_alias` and the `MtyL_transparent` cases.

samsa1 commented on typing/includemod.ml:
```diff
@@ -430,7 +430,12 @@ let retrieve_functor_params env mty =
+        | Ok mty ->  retrieve_functor_params before env mty
+        | Error _ -> { Error.params = List.rev before; res }
+        end
+    | Mty_transparent p as res ->
```
This looks like code duplication for the `Mty_static_alias` and the `Mty_transparent` cases

samsa1 commented on typing/includemod_errorprinter.ml:
```diff
@@ -154,15 +154,15 @@ module Runtime_coercion = struct
      a [pos] path *)
   let rec find env ctx path (mt:Types.module_type) = match mt, path with
-    | (Mty_ident p | Mty_alias p), _ ->
+    | (Mty_ident p | Mty_static_alias p), _ ->
```
Isn't a `Mty_transparent` missing here ? If not maybe adding a case at the end explaining why it should go in the default case might be usefull

samsa1 commented on typing/mtype.ml:
```diff
@@ -189,41 +189,44 @@ let rec nondep_mty_with_presence env va ids pres mty =
+          in
+          nondep_mty env va ids expansion.md_type
+      | None -> mty
+      end
```
This looks like code duplication between the `Mty_static_alias` and the `Mty_transparent` cases

samsa1 reviewed (commented):
Did a first global pass and found a few times things that looked like code duplication. Is it because you expect in the future that both cases will not behave the same ? Or did I miss differences between the way both cases are handled ?

(I haven't looked at code correctness yet)

OlivierNicole commented on lambda/translmod.ml:
```diff
@@ -930,20 +929,20 @@ and all_idents = function
     | Tstr_include incl ->
       bound_value_identifiers incl.incl_type @ all_idents rem
 
+    | Tstr_module {mb_id = None} -> all_idents rem
+    | Tstr_module {mb_expr = {mod_type}} when (Types.mty_is_absent mod_type) ->
+        all_idents rem
```
This could be an or-pattern

OlivierNicole commented on typing/env.mli:
```diff
@@ -375,8 +375,7 @@ val enter_extension:
-  scope:int -> ?noalias:bool -> string ->
-  module_type -> t -> Ident.t * t
+  scope:int -> ?noalias:bool -> string -> module_type -> t -> Ident.t * t
 val enter_module_declaration:
```
This formatting change seems unrelated to the present commit.

OlivierNicole commented on typing/mtype.ml:
```diff
@@ -354,10 +349,10 @@ and no_code_needed_sig env sg =
       | Val_prim _ -> no_code_needed_sig env rem
       | _ -> false
       end
-  | Sig_module(id, _, ({md_type = Mty_static_alias _} as md), _, _) :: rem ->
```
Nit: the new formatting of this record pattern seems somewhat inconsistent.

OlivierNicole commented on typing/typemod.ml:
```diff
@@ -2405,61 +2405,52 @@ let rec type_module ?(alias=false) ~strengthen ~funct_body anchor env smod =
+                                 mod_loc = smod.pmod_loc } in
+      let aliasable = Env.is_aliasable path env in
+      (* Error case *)
+      if (static_alias_flag || dynamic_alias_flag) && (not aliasable) then
+        (* Non aliasable paths are rejected *)
+        raise (Error (smod.pmod_loc, env, Cannot_alias path));
```
This error doesn’t seem to be raised in `type_module_aux` on trunk. However, this kind of invalid alias errors are caught nonetheless. Could you help me understand what is the code path that catches this `Cannot_alias` error on current trunk? It would probably help me check that this change does what is expected.

OlivierNicole reviewed (commented):
Thank you for this very clear and pleasant to review PR. Overall I believe that the changes behave as stated, save for a few questions below.

I also have two questions/remarks that are not about a specific code change:

- I’m not sure to understand why the bootstrap commits change `boot/lexer.ml`, even though no new syntax is introduced.
- There is a typo in commit “Remove presence computation in mtype”: has being -> as being

clementblaudeau committed:
fix: map the long ident with [map_loc_lid] instead of [map_loc]

clementblaudeau committed:
test: add a passing test for dynamic aliases

clementblaudeau committed:
refactor: catch invalid alias earlier

clementblaudeau committed:
refactor: remove [make_alias_absent]

The mechanism for making alias absent is no longer needed, as dynamic aliases
are never turned into static ones. A side effect is that [module type of], using
strengthening, creates dynamic aliases, while inference still produces static
aliases by default. This prevents patterns like:

```ocaml
module X0 = struct
  module X1 = struct end
end
module Y0 = X0
module M : module type of Y0.X1 = X0.X1
(* Error: Signature mismatch:
       Modules do not match:
         (module X0.X1) [@static_alias]
       is not included in
         (module X0.X1) [@dynamic_alias] *)
```

This can be fixed by explicitly requiring a dynamic alias (which will eventually
be the default behavior of inference):

```ocaml
module M : module type of Y0.X1 = X0.X1 [@dynamic_alias]
```

The test gatien_baron_20131019_ok.ml was changed for that reason.
